### PR TITLE
Move book examples into standalone projects. Include them in book via mdbook preprocessor.

### LINKED ISSUE #544: Move docs code examples to real projects
Whole-code examples in the docs should be moved to a directory with full projects (and potentially tests), and added to CI to at least build. Then the code can be included using [`{{ #include }}`](https://rust-lang.github.io/mdBook/format/configuration/preprocessors.html).

### PR DESCRIPTION
**Pull request by mitchmindtree**
Closes #544, follows the example set in #731.

This also adds all examples to the Sway workspace so that we can test both projects under CI. Doing so means the `Cargo.lock` for the workspace requires an update, but I haven't included it yet to reduce noise in the PR. @adlerjohn shall I add the `Cargo.lock` changes as a final commit in this PR? Or do you generally update the lock file separately? 

## TODO

- [x] `hello_world`
- [x] `fizzbuzz`
- [x] `wallet_smart_contract`
- [x] Ensure tests are run for examples in CI.
- [x] Ensure `forc build` is ran for each example in CI.

### CODE REVIEW

File: docs/src/getting-started/forc_project.md [hunk 1]
```diff
@@ -39,7 +31,7 @@ Here are the contents of the only Sway file in the project, and the main entry p
 script;
 
 fn main() {
-
```
Comment (mitchmindtree) [inline docs/src/getting-started/forc_project.md:42]: Woops, my editor's setup up to remove trailing whitespace automatically when writing the buffer

File: examples/hello_world/tests/harness.rs [hunk 2]
```diff
@@ -0,0 +1,43 @@
+use fuel_tx::Salt;
+use fuels_abigen_macro::abigen;
+use fuels_contract::contract::Contract;
+use rand::rngs::StdRng;
+use rand::{Rng, SeedableRng};
+
+// Generate Rust bindings from our contract JSON ABI
+// FIXME: Incorrect path, see https://github.com/FuelLabs/fuels-rs/issues/94
+abigen!(MyContract, "./examples/hello_world/my-contract-abi.json");
```
Comment (mitchmindtree) [inline examples/hello_world/tests/harness.rs:9]: We might need to resolve https://github.com/FuelLabs/fuels-rs/issues/94 before landing this, as this path might be confusing for newcomers reading through the book.

**Comment by adlerjohn** [comment]
If you're going to update the workspace or Manifest file deps, then you can update the lockfile in the same PR. _That being said_, I'm actually against putting the examples in the workspace _because_ that would modify the lockfile. We had to [remove examples from `fuel-core`](https://github.com/FuelLabs/fuel-core/pull/140) _and move them to a whole new repo_ to avoid dependency issues. My preference would be instead to add a new task to the CI pipeline to go into each example sub-directory and test it.

**Comment by mitchmindtree** [comment]
> We had to https://github.com/FuelLabs/fuel-core/pull/140 and move them to a whole new repo to avoid dependency issues.

Aghhhhh I see, that makes sense. Things get tricky when you build the thing you use to build things in the same repo :joy: 

> My preference would be instead to add a new task to the CI pipeline to go into each example sub-directory and test it.

Sounds good - I'm in the middle of writing a little rust "script" that runs `forc build` for each forc project under the `examples` directory solely for the purpose of running in CI. I'll try modifying it to also run the tests manually for each so we can remove them from the workspace altogether.

File: examples/build-all-examples/Cargo.toml [hunk 3]
```diff
@@ -0,0 +1,6 @@
+[package]
+name = "build-all-examples"
+description = "Runs `forc build` for all projects under the Sway `examples` directory."
+version = "0.1.0"
```
Comment (adlerjohn) [inline examples/build-all-examples/Cargo.toml:4]:
```suggestion
version = "0.0.0"
```

File: Cargo.lock [hunk 4]
```diff
@@ -937,7 +941,7 @@ checksum = "3f9eec918d3f24069decb9af1554cad7c880e2da24a9afd88aca000531ab82c1"
 
 [[package]]
 name = "forc"
-version = "0.3.3"
+version = "0.4.0"
```
Comment (adlerjohn) [inline Cargo.lock:944]: Ahh, could you make a separate PR to bump the version of these in the lockfile? Looks like #760 forgot the bump the lockfile.

File: Cargo.toml [hunk 5]
```diff
@@ -2,6 +2,7 @@
 resolver = "2"
 members = [
     "docstrings",
+    "examples/build-all-examples",
```
Comment (adlerjohn) [inline Cargo.toml:5]: Hmmm. Should this not be in the workspace?

Comment (mitchmindtree) [reply Cargo.toml:5]: This should be fine - the `build-all-examples` "script" itself has no direct crate dependencies so shouldn't cause any conflicts. Actually *running* the script *does* expect there to be an instance of `forc` available in `PATH`, but this doesn't affect the workspace.

The idea behind `build-all-examples` is something like this: After all the existing CI passes, we `cargo install --path ./forc` so that its available via `PATH`. Then we `cargo run --bin build-all-examples` which will use the installed `forc` instance to build each example individually.

Comment (adlerjohn) [reply Cargo.toml:5]: Ah I see, and `cargo run --bin build-all-examples` requires `build-all-examples` to be built, which is done automagically if it's in the workspace, right?

Comment (mitchmindtree) [reply Cargo.toml:5]: Yeah exactly.

Looks like it's working! We can see the "Build all examples" CI step fail [here](https://github.com/FuelLabs/sway/runs/5142599565?check_suite_focus=true#step:13:101), where `forc build` fails to compile the `wallet_smart_contract` example as `storage` isn't yet implemented. Hopefully this helps to keep the book examples up to date.

Next step is to work out how to run `forc test` for each example too. This is a bit trickier as, unlike `forc build`, `forc test` invokes `cargo test` under the hood and in this case it's happening *inside* the sway workspace, so we need to find a way to get this cargo invocation to "ignore" the sway workspace. Will have a look at this tomorrow!

**Comment by mitchmindtree** [comment]
OK, I think this is just about ready to go!

**Testing `examples`**

I think it's best to leave the example *testing* (i.e. running `forc test` for each) to a future PR as there are a few issues involved that should be worked through first. I've opened #771 to track this.

**`wallet_smart_contract` failure**

Currently the `wallet_smart_contract` example fails during `forc build` as `storage` isn't yet implemented. We could leave this PR up until `storage` lands (it sounds like its not far off?) or add a hack to ignore this example for now?

**Comment by mitchmindtree** [comment]
Okydoke, I had to comment out a large portion of the wallet example, but I've left a `NOTE` at the top with a link to the PR. Once this lands I'll open a dedicated issue for updating the `wallet` example.

**Review by sezna — CHANGES_REQUESTED** [review]
Generally LGTM, left some nits

**Comment by mitchmindtree** [comment]
@sezna good catch, should be addressed now!

**Review by adlerjohn — APPROVED** [review]

**Review by sezna — APPROVED** [review]
